go-controller/pkg/ovn/policy_retry.go

```diff
@@ -0,0 +1,186 @@
+package ovn
+
+import (
+	"fmt"
+	knet "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"math/rand"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+// iterateRetryNetworkPolicies checks if any outstanding NetworkPolicies exist
+// then tries to re-add them if so
+// updateAll forces all policies to be attempted to be retried regardless
+func (oc *Controller) iterateRetryNetworkPolicies(updateAll bool) {
+	oc.retryNetPolLock.Lock()
+	defer oc.retryNetPolLock.Unlock()
+	now := time.Now()
+	for namespacedName, npEntry := range oc.retryNetPolices {
+		if npEntry.ignore {
+			continue
+		}
+
+		// check if we need to create
+		var policyToCreate *knet.NetworkPolicy
+		if npEntry.newPolicy != nil {
+			// get the latest version of the new policy from the informer, if it doesn't exist we are not going to
+			// create the new policy
+			np, err := oc.watchFactory.GetNetworkPolicy(npEntry.newPolicy.Namespace, npEntry.newPolicy.Name)
+			if err != nil && errors.IsNotFound(err) {
+				klog.Infof("%s policy not found in the informers cache, not going to retry policy create", namespacedName)
+				npEntry.newPolicy = nil
+			} else {
+				policyToCreate = np
+			}
+		}
+
+		npEntry.backoffSec = npEntry.backoffSec * 2
+		if npEntry.backoffSec > 60 {
+			npEntry.backoffSec = 60
+		}
+		backoff := (npEntry.backoffSec * time.Second) + (time.Duration(rand.Intn(500)) * time.Millisecond)
+		npTimer := npEntry.timeStamp.Add(backoff)
+		if updateAll || now.After(npTimer) {
+			klog.Infof("Network Policy Retry: %s retry network policy setup", namespacedName)
+
+			// check if we need to delete anything
+			if npEntry.oldPolicy != nil {
+				klog.Infof("Network Policy Retry: Removing old policy for %s", namespacedName)
+				if err := oc.deleteNetworkPolicy(npEntry.oldPolicy, npEntry.np); err != nil {
```

Comment:
Is this the only spot where we actually use `npEntry.np` value? Oh its for the case where we have deleted the namespace so no entry in nsInfo but policy is still stuck on retry...
q: does k8s have these finalizer thing set on namespace wherein deletion of namespace is stuck on objects inside it going away? feels weird that namespace is gone and there are objects pointing to it, but that's in the k8s world. I guess your fix here in ovn-k makes sense.

Reply:
I'm not 100% sure, but yeah I think in KAPI a deletion of a parent object can only occur with deletion of all references to it. However, on our side what if our code races? We could have namespace delete object and policy delete object enter at the same time, namespace handler attempts to delete the np, somehow fails and then has to remove the namespace from the cache. Now policy handler cannot get a lock on nsInfo to find the np. Well now this problem is solved by this PR :)